accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,354 @@
+    {
+        // This compiles today:
+        ref Span<int> s = ref __refvalue(tr, Span<int>);
+
+        // And can be assigned to default
+        s = default;
+
+        Span<int> newspan = stackalloc int[42];
+
+        // But this causes CS8352 as expected:
+        s = ref newspan;
+    }
+
+    public static void CallMe(ref Span<int> span1, Span<int> span2) { }
+}
+
+public ref struct MyRefStruct
+{
+    public Span<int> _span;
+}
+```
+
+If allowed today in strongly-typed manner, then reflection should also allow:
+```cs
+internal class Program
+{
+    static void Main()
+    {
+
+        Span<int> span = stackalloc int[42];
+        TypedReference tr = __makeref(span); // Assume we allow instead of CS1601
+
+        // Using a proposed invoke API; calling should be supported passing byvalue
+        MethodInfo mi1 = typeof(Program).GetMethod(nameof(ChangeIt1));
+        mi1.InvokeDirect(target: default, arg1: tr);
+
+        // and supported passing byref
+        MethodInfo mi2 = typeof(Program).GetMethod(nameof(ChangeIt2));
+        mi2.InvokeDirect(target: default, arg1: tr);
+
+        // Just like these methods can be called today:
+        ChangeIt1(span);
+        ChangeIt2(ref span);
+    }
+
+    public static void ChangeIt1(Span<int> span) { }
+    public static void ChangeIt2(ref Span<int> span) { }
+}
+```
+
+FWIW `__arglist` currently compiles with byref-like types:
+```cs
+Span<byte> s1 = stackalloc byte[1];
+Span<byte> s2 = stackalloc byte[11];
+Span<byte> s3 = stackalloc byte[111];
+CallMe(__arglist(s1, s2, s3));
+
+static unsafe void CallMe(__arglist)
+{
+    // However, when enumerating __arglist here, the runtime throws when accessing a byref-like
+    // type although that limitation could be removed (just a runtime limitation; not compiler)
+```
+
+## Variable-length, safe collections
+(todo; see https://github.com/dotnet/runtime/issues/75349. _I have a local prototype that does a callback that doesn't require any new language or runtime features however it is not a clean programming model. The callback is necessary to interop with hide our internal GC "tracking" feature via "RegisterForGCReporting()"._)
+
+## Exception handling
+Todo; discuss using `BindingFlags.DoNotWrapExceptions` semantics only (no wrapping of exceptions).
+
+# Proposed APIs
+(work in progress; various prototypes exist here)
+## TypedReference
+```diff
+namespace System
+{
+    public ref struct TypedReference
+    {
+        // Equivalent of __makeref (except for a byref-like type since they can't be a generic parameter)
++       public static TypedReference Make<T>(ref T? value);
+        // Helper used for boxed or loosely-typed cases
++       public static TypedReference Make(ref object value, Type type);
+
+        // Equivalent of __refvalue
++       public ref T GetValue<T>();
+        // Used for byref-like types or loosely-typed cases with >4 parameters
++       public readonly unsafe ref byte TargetRef { get; }
```

Comment:
This is for an on stack array, right?

Reply:
Yes it would be used to help "peel off" the reference so it can be passed as a non-Typed referenced. See this [prototype ](https://github.com/dotnet/runtime/compare/main...steveharter:runtime:InvokeWithStj#diff-a6a6bf9694f968751e223700b17bb8588bddc4a9ed62e0a9d8534bf390e60891R68) for the simple case of a single parameter (property value).